client/code_generator/generator.go

```diff
@@ -1,10 +1,295 @@
+		color.Aurora.BrightYellow(reflect.TypeOf(structure).Name()),
+	))
+}
+
+func (g *generator) generateAttributeUpdate(au model.AttributeUpdate) {
+	switch partUpdate := au.PartUpdate.(type) {
+	case model.OctaveSet:
+		g.addToken(
+			parser.OctaveSet, fmt.Sprintf("o%d", partUpdate.OctaveNumber),
+		)
+	case model.OctaveUp:
+		g.addToken(parser.OctaveUp, ">")
+	case model.OctaveDown:
+		g.addToken(parser.OctaveDown, "<")
+	default:
+		// No other AttributeUpdates can be introduced directly into Alda
+		g.logError(partUpdate)
+	}
+}
+
+func (g *generator) generateGlobalAttributeUpdate(
+	ga model.GlobalAttributeUpdate,
+) {
+	// GlobalAttributeUpdates can only be introduced using Lisp
+	g.logError(ga)
+}
+
+func (g *generator) generateBarline(_ model.Barline) {
+	g.addToken(parser.Barline, "|")
+}
+
+func (g *generator) generateChord(chord model.Chord) {
+	hasNoteOrRestBeenEncountered := false
+
+	for _, update := range chord.Events {
+		updateType := reflect.TypeOf(update)
+		isNoteOrRest := updateType == reflect.TypeOf(model.Note{}) ||
+			updateType == reflect.TypeOf(model.Rest{})
+
+		if isNoteOrRest && hasNoteOrRestBeenEncountered {
+			g.addToken(parser.Separator, "/")
+		}
+
+		if isNoteOrRest {
+			hasNoteOrRestBeenEncountered = true
+		}
+
+		g.generateTokens([]model.ScoreUpdate{update})
+	}
+}
+
+func (g *generator) generateCram(cram model.Cram) {
+	g.addToken(parser.CramOpen, "{")
+	g.generateTokens(cram.Events)
+	g.addToken(parser.CramClose, "}")
+	g.generateDuration(cram.Duration)
+}
+
+func (g *generator) generateEventSequence(es model.EventSequence) {
+	g.addToken(parser.EventSeqOpen, "[")
+	g.generateTokens(es.Events)
+	g.addToken(parser.EventSeqClose, "]")
+}
+
+func (g *generator) generateLispNil(n model.LispNil) {
+	g.logError(n)
+}
+
+func (g *generator) generateLispList(l model.LispList) {
+	// The MusicXML importer currently has no use for Lisp
+	g.logError(l)
+}
+
+func (g *generator) generateMarker(marker model.Marker) {
+	g.addToken(parser.Marker, "%" + marker.Name)
+}
+
+func (g *generator) generateAtMarker(atMarker model.AtMarker) {
+	g.addToken(parser.AtMarker, "@" + atMarker.Name)
+}
+
+func (g *generator) generatePitch(pitch model.PitchIdentifier) {
+	switch typedPitch := pitch.(type) {
+	case model.LetterAndAccidentals:
+		g.addToken(
+			parser.NoteLetter,
+			strings.ToLower(typedPitch.NoteLetter.String()),
+		)
+
+		for _, accidental := range typedPitch.Accidentals {
+			switch accidental {
+			case model.Flat:
+				g.addToken(parser.Flat, "-")
+			case model.Natural:
+				g.addToken(parser.Natural, "_")
+			case model.Sharp:
+				g.addToken(parser.Sharp, "+")
+			}
+		}
+	default:
+		// MidiNoteNumber has no direct representation in Alda syntax, and is
+		// only ever obtained from Lisp
+		// The MusicXML parser uses MidiNoteNumber pitches for percussion notes
+		// but translates this to LetterAndAccidentals in the optimization phase
+		g.logError(pitch)
```

Comment:
I think it might be reasonable in this case for the code generator to generate a Lisp AST node like `(note (pitch (midi-note 42)))`

Reply:
So far I've chosen to not deal with any Lisp within the code generator, simply because the MusicXML importer will never generate it. In fact there's a few more things I've chosen to ignore as well. But if we're going to go towards having code generation be a `ScoreUpdate` interface, now we should consider how to approach all the methods I left unhandled. 

The problem here is I'm not really sure how to handle most of those `ScoreUpdate` elements. Most of them have no equivalent syntactic representation in Alda code, and can only be parsed to `ScoreUpdate` via Lisp and then possibly evaluating that Lisp. 

## Dealing with `ScoreUpdate`s
So one option is to somehow just generate Lisp from all of these `ScoreUpdate`s that don't have corresponding syntax. This would be hard, and I don't understand enough about Alda to even be sure if this works. **My intuition is it's possible to handle code generation for any arbitrary Alda `ScoreUpdate`**. 

For example, what if there's a note with a nice letter and accidentals pitch, a nice note length ms duration, but also a note length beats duration? This is valid Alda internal representation. But the only way to generate this would be to generate the entire `ScoreUpdate` as Lisp. This will cause problems for sure. 

So `ScoreUpdate` is fundamentally a superset of Alda source code. And if we try to embed code generation as a `ScoreUpdate` interface, there will be times we have to do weird workarounds or return error. I'm not as familiar with Alda as I probably should be to talk about this, but this is the understanding I have so far. 

## Keeping things the same
Then the other option is to simply leave it as is. We don't spin off code generation as a `ScoreUpdate` interface. We keep all of the logic in one place so we can easily see what is handled and what isn't. I think there's an argument for this. 

Code generation will only be used when importing from external sources. And when someone is writing that importer, they will be putting in a lot of work, which involves testing if it can be completely decompiled into source code. In this process they will have to look at the code generator to make sure everything they build can be generated. Then support can be added as necessary. 

Another option is to really make this distinction clear. We can put `code_generator` in the `interop` folder. Then `code_formatter` in the base `client` folder. This is because formatting can be done on Alda in general, and is closer to the language itself. But generation is only on imported representations which will happen in the `interop` folder.

Reply:
I had a glance over all of the various things that implement the ScoreUpdate interface, and I believe it should be possible for us to add `AST()` as an interface method. There are a couple of interesting ones, like:

* AttributeUpdate
  * Has a PartUpdate, so we could also add `AST()` to the PartUpdate interface and have the compiler force us to define how each PartUpdate would be represented e.g. in the inline Lisp. I would view this as a positive thing, because it would encourage us to make sure that all kinds of part updates are supported in Lisp.
* GlobalAttributeUpdate
  * Also has a PartUpdate, but the only difference is that we need to generate e.g. `volume!` instead of `volume`. I guess one way we could do this is to add another method to the PartUpdate interface to explicitly define how the _global_ version of that part update would be represented as an AST. That feels kind of gross, but I think I'd be willing to accept the grossness as a trade-off for the benefits of this approach where the compiler helps us make sure that everything has an AST representation.
* LispNil, LispList
  * These are the only two things in lisp.go that implement ScoreUpdate, and they would be trivial to represent in the AST, because of the one-to-one relationship of Lisp code to the AST. When we implement the AST, I think we'll probably end up just having a corresponding AST node type for each kind of value possible in the Lisp, e.g. an AST LispString that corresponds to model.LispString, and the same thing for LispNumber, LispList, etc. etc.
* Note, Rest
  * These will be a bit tricky, as you mentioned, but not _too_ tricky, I don't think.
  * We can special-case notes where the pitch is represented as a MIDI note number and choose to return the note in Lisp form.
  * As for the thing about durations: I don't think it's possible for a Note to have a duration expressed in both beats and milliseconds simultaneously. A Note has a Duration, which has a list of DurationComponents. DurationComponent is an interface with specific implementations including NoteLength, NoteLengthBeats, and NoteLengthMs. We could add an `AST()` function to the DurationComponent interface as well, so that the compiler ensures that we always define how to generate the AST for these things. If there are multiple duration components, then we tie them together with a tie (`~`) node in between each component.